pydantic_source/source.py

```diff
@@ -0,0 +1,158 @@
+import json
+import ast
+from typing import Any, Type, Dict
+from benedict import benedict
+from munch import Munch
+
+import yaml
+from pydantic_settings import BaseSettings, PydanticBaseSettingsSource
+from pydantic.fields import FieldInfo
+
+from rapyuta_io_sdk_v2 import Client, Configuration
+import base64
+
+
+class ConfigTreeSource(PydanticBaseSettingsSource):
+    def __init__(
+        self,
+        settings_cls: Type["BaseSettings"],
+        config: Configuration,
+        tree_name: str = "default",
+        key_prefix: str = "",
+        local_file: str = None,
+    ):
+        super().__init__(settings_cls)
+        self.client = Client(config=config)
+        self.tree_name = tree_name
+        self.local_file = local_file
+        self._top_prefix = key_prefix
+
+        # ? Placeholder for the configuration tree data
+        self.config_tree = None
+
+        # * Load the configuration tree
+        self.load_config_tree()
+
+        processed_data = self._process_config_tree(raw_data=self.config_tree)
+
+        if processed_data is None:
+            raise ValueError("processed_data cannot be None")
+
+        self._configtree_data = benedict(processed_data).unflatten(separator="/")
+
+    # * Methods to fetch Configtree
+    def fetch_from_api(self):
+        """
+        Load the configuration tree from an external API.
+        """
+        try:
+            response = self.client.get_configtree(
+                name=self.tree_name,
+                include_data=True,
+                content_types="kv",
+                with_project=False,
+            )
+            config_tree_response = Munch.toDict(response).get("keys")
+            self.config_tree = self._extract_data_api(input_data=config_tree_response)
+        except Exception as e:
+            raise ValueError(f"Failed to fetch configuration tree from API: {e}")
+
+    def load_from_local_file(self):
+        """
+        Load the configuration tree from a local JSON or YAML file.
+        """
+        if not self.local_file:
+            raise ValueError("No local file path provided for configuration tree.")
+
+        try:
+            with open(self.local_file, "r") as file:
+                if self.local_file.endswith(".json"):
+                    self.config_tree = self._extract_data_local(json.load(file))
+                    self.config_tree = benedict(self.config_tree).flatten(separator="/")
+                elif self.local_file.endswith(".yaml") or self.local_file.endswith(
+                    ".yml"
+                ):
+                    self.config_tree = self._extract_data_local(yaml.safe_load(file))
+                    self.config_tree = benedict(self.config_tree).flatten(separator="/")
```

Comment:
todo: Can be simplified by just passing the format to benedict. See the _process_files_with_override in CLI.